ote/src/clj/ote/services/users.clj

```diff
@@ -0,0 +1,272 @@
+            [specql.op :as op]
+            [ote.db.user :as user]
+            [ote.services.login :as login]
+            [ote.db.auditlog :as auditlog]
+            [ote.db.tx :as tx :refer [with-transaction]]
+            [ote.localization :as localization :refer [tr]]
+            [ote.util.encrypt :as encrypt]
+            [specql.core :as specql]
+            [clojure.set :as set]
+            [clojure.string :as str]
+            [buddy.hashers :as hashers]
+            [ote.localization :as localization]
+            [ote.email :as email]
+            [ote.util.email-template :as email-template]
+            [hiccup.core :as hiccup]
+            [taoensso.timbre :as log]
+            [ote.time :as time]
+            [clj-time.core :as t]
+            [clj-time.coerce :as tc]
+            [ote.util.throttle :as throttle]
+            [clj-time.coerce :as tc])
+  (:import (java.util UUID)))
+
+(defqueries "ote/services/user_service.sql")
+
+(defn delete-users-old-token!
+  [db user-email]
+  (specql/delete! db ::user/email-confirmation-token
+    {::user/user-email user-email}))
+
+(defn send-email-verification
+  [email-config user-email language email-confirmation-uuid]
+  (try
+    (localization/with-language
+      language
+      (email/send!
+        email-config
+        {:to user-email
+         :subject (tr [:email-templates :email-verification :verify-email])
+         :body [{:type "text/html;charset=utf-8"
+                 :content (str email-template/html-header
+                            (hiccup/html (email-template/email-confirmation (tr [:email-templates :email-verification :verify-email]) email-confirmation-uuid)))}]}))
+    (catch Exception e
+      (log/warn (str "Error while sending verification to:  " user-email " ") e))))
+
+(defn create-confirmation-token!
+  [db user-email token]
+  (let [expiration-date (time/sql-date (.plusDays (java.time.LocalDate/now) 7))]
+    (specql/insert! db ::user/email-confirmation-token
+      {::user/user-email user-email ::user/token token ::user/expiration expiration-date})))
+
+(defn valid-user-save? [{:keys [username name email]}]
+  (and (user/email-valid? email)
+    (user/username-valid? username)
+    (string? name)
+    (not (str/blank? name))))
+
+(defn save-user!
+  [db email-config user form-data admin?]
+  (if-not (valid-user-save? form-data)
+    {:success? false}
+    (with-transaction db
+      (let [{new-email :email
+             new-username :username} form-data
+            username-taken? (and (not= (:username user) new-username)
+                              (username-exists? db {:username new-username}))
+            language (or (:language form-data) :fi)
+            email-changed? (not= (:email user) new-email)
+            email-taken? (and email-changed?
+                           (email-exists? db {:email new-email}))
+            login-info (first (login/fetch-login-info db {:email (:email user)}))
+            password-incorrect? (if admin?
```

Comment:
A matter of taste, & this was already discussed, in any case making a note:
negations and inversions should be avoided in naming for better clarity.
```suggestion
            password-correct? (if admin?
                                             ( ...code...)
                                             false)
```